### DISCUSSION TITLE
Version 1.9 throwing errors where version 1.4 was not

### DISCUSSION THREAD

**Discussion by roarengineering** [reduxjs/redux-toolkit#3416] [Q&A]
I recently updated my project from version 1.4.0 to version 1.9.2 and have been seeing this error in various places  since then:

cannot assign to read only property "xxxx" of object "[]<object>"

Nothing changed besides the upgrade. I want to better understand the error, but also what changed between versions and why, so that I can decide the best way to approach fixing this.

Any explanation for the loss of backwards compatibility would be great!

Thanks

**Comment by EskiMojo14** [comment]
Immer freezes the state it produces, to prevent accidental mutation. It sounds like you're trying to mutate data you're getting from the store, which you shouldn't ever do.

1.4.0 was a lot of versions ago, so I'm not sure what change in particular introduced this.

**Comment by roarengineering** [comment]
It seems it may have been introduced with this version, according to the release notes.

<img width="389" alt="image" src="https://user-images.githubusercontent.com/126616399/236095741-05deca02-33cd-4e2a-8362-c4cfa745e847.png">

**Reply by markerikson** [reply]
Yep, that's probably it.

To be clear, most likely your existing code _was buggy_.  Immer auto-freezing state is highlighting the existing bug in the app code.

**Reply by roarengineering** [reply]
Do you know of a linter that can help me find other areas where this happens? Since this doesn't cause build issues I'd have to test everything.

**Reply by phryneas** [reply]
That error message pretty much *is* the linter - somewhere in your component code, you are modifying an object that exists in Redux state, also modifying your Redux state. Without this error, that means that your component would internally change, but not rerender - and cause all kinds of "whacky" behavior.

But you probably won't get around testing your app by hand to trigger those warnings and find where you had that bug, I'm sorry :/
